docs/kb/semgrep-ci/scan-monorepo-in-parts.md

```diff
@@ -0,0 +1,54 @@
+---
+description: How to scan a monorepo in parts for better CI performance and clearer findings organization
+tags:
+  - monorepo
+  - optimization
+  - performance
+  - ci
+---
+
+# Scanning a monorepo in parts
+
+Monorepos, by definition, contain a large amount of code. The code is also often divided into to different components or modules, despite all existing in the same repository.
+
+As such, it can be helpful to scan a monorepo in parts for multiple reasons:
+
+1. Improving scan performance in CI and reducing CI run times
+2. Logically splitting up the monorepo into multiple components to help with managing findings
+
+## How to configure Semgrep in CI to split up a monorepo
+
+Typically when scanning any repo with Semgrep in CI, the base command is `semgrep ci`. Of course, you need to have an auth token set and perhaps other environment variables configured depending on what SCM and CI provider you are using, but this article will assume you are already familiar with those settings.
+
+To split up your monorepo, you need to first decide *how* you want to logically split up the code. For example, if your monorepo has four main modules and their paths are:
+
+    /src/moduleA
+    /src/moduleB
+    /src/moduleC
+    /src/moduleD
+
+Then it would logically make sense to split this into four separate scans in CI, one for each module.
+
+To actually achieve this, you would use the `--include` flag ([see CLI reference](/docs/cli-reference/)) to only scan files in each module's code path:
+
+    semgrep ci --include=/src/moduleA/*
+
+Now, Semgrep is only scanning files under that path and the entire CI run will take less time, since less code is being scanned. Of course, you will now have to run multiple CI runs for the remaining three modules to get full monorepo coverage, but you have more flexibility on how and when you perform those runs.
+
+Now that you've successfully configured your monorepo to be scanned in parts, you also have to configure the findings from each part or module to show up as their own project in Semgrep Cloud Platform (SCP).
```

Comment:
I'd still suggest integrating this further, since we talk about separating findings as part of the benefit at the very top. Maybe something like changing line 32 above to:

> After choosing a logical split, use the `--include` flag ([see CLI reference](/docs/cli-reference/)) with the relevant path to only scan files in that module's code path, and add the the `SEMGREP_REPO_NAME` environment variable ([see CI environment variables reference](/docs/semgrep-ci/ci-environment-variables/#semgrep_repo_name)) to send the findings to a different project in Semgrep.

Reply:
I do get the idea behind this, but I also think it muddies up the 2 concepts, 1) being dividing up the paths with --include and 2) being sending the findings to separate projects. I'm afraid if I write something similar to your suggestion, it will distract from just getting part 1) right. It makes sense in my mind to have them as 2 separate parts but emphasize that both are needed for this whole thing to work.